nbviewer/providers/s3/handlers.py

```diff
@@ -0,0 +1,149 @@
+# -----------------------------------------------------------------------------
+#  Copyright (C) Jupyter Development Team
+#
+#  Distributed under the terms of the BSD License.  The full license is in
+#  the file COPYING, distributed as part of this software.
+# -----------------------------------------------------------------------------
+import errno
+import io
+import os
+from datetime import datetime
+from urllib.parse import urlparse
+
+import boto3
+import botocore
+from tornado import iostream
+from tornado import web
+
+from .. import _load_handler_from_location
+from ...utils import url_path_join
+from ..base import cached
+from ..base import RenderingHandler
+
+
+class S3Handler(RenderingHandler):
+    """Renderer for s3://
+
+    Serving notebooks from S3 buckets
+    """
+
+    def initialize(self, **kwargs):
+        self.s3_client = boto3.client("s3")
+        self._downloadable_data = None
+        self._downloaded_path = None
+        super().initialize(**kwargs)
+
+    async def download(self, path):
+        """Download the notebook"""
+        headers = await self.get_notebook_headers(path)
+        filename = os.path.basename(path)
+        self.set_header("Content-Length", headers["ContentLength"])
+        # Escape commas to workaround Chrome issue with commas in download filenames
+        self.set_header(
+            "Content-Disposition",
+            "attachment; filename={};".format(filename.replace(",", "_")),
+        )
+        if self._downloaded_path == path and self._downloadable_data is not None:
```

Comment:
I wanted to prevent re-fetching the file from s3. Maybe there is a beter way to cache the boto3 s3 transfer using memcached somehow. Any suggestions are welcome.